Refactored method GenerateStringConcat in Parser.cs

nemo048 opened the pull request:
Fixes #151 

It would be great if you could review my changes for Parser.GenerateStringConcat method in this pull request.

PS. I also added two static fields for MethodInfos you use in there so that they won't have to be calculated on each call

davideicardi commented:
Great! Thank you!

Can you just provide an unit test to verify that? Maybe concateneting a string with some other object? Or do some similar test?
Potentially you can also test the generated expression. What do you think?

nemo048 commented:
Should I add a reference to Entity Framework Core to unit tests project or I just need to test the existing method?

Also there is already a couple of tests with concatenation of two strings and a string with an integer. Do you need anything more than this?

davideicardi commented:
No, I don't think that we have to reference Entity Framework. But in some way we should test the expected behaviour.

Maybe you can just verify that calling `Concat`  will result in calling `String.Concat(String, String)` instead of `String.Concat(Object, Object)`? I think that you can do that by inspecting the resulted expression. Right?

Also what happen if we have a `null`? It is supported?

nemo048 commented:
I've added two more tests: one is asserting that we call `String.Concat(String, String)` method and another one is checking null to string concatenation behaviour. And yes, now null is supported 😄

nemo048 commented:
Sorry 😄  Could you please say is it enough to merge or should I do something more?

davideicardi commented on test/DynamicExpresso.UnitTest/OperatorsTest.cs:
```diff
@@ -130,6 +133,34 @@ public void String_Concatenation_with_type_conversion()
+
+            Assert.AreEqual("ciao ", interpreter.Eval(expressionText));
+        }
+
```
What if you write:
```
        [Test]
        public void String_Concatenation_with_null()
        {
            Interpreter interpreter = new Interpreter();
            string expressionTextWithNullKeyword = "\"ciao \" + null";
            Assert.AreEqual("ciao ", interpreter.Eval(expressionTextWithNullKeyword));

            // create a function that returns null
            Func<String> someFunc = () => null;
            interpreter.SetFunction("someFunc", someFunc);
            string expressionTextWithNullFunc = "\"ciao \" + someFunc()";
            Assert.AreEqual("ciao ", interpreter.Eval(expressionTextWithNullFunc));
        }
```
My concern is that we should not call `.ToString` directly, because on null it will not work. But checking if the expression is null I think is not enough.
Maybe there is some better solution, but I need to think about it.

Also, one minor fix, sorry ;-) : can you please align code space style?
The repo is configured using tabs, I think that you have used spaces.

davideicardi reviewed (changes requested):
Some considerations about null expression, and one small indentation problem.
Thanks!

metoule commented:
It might also be a good idea to add a test with a variable of a type that overrides the `ToString` method, to ensure the overridden method is called (and not the base implementation). Something like this:

```csharp
public class MyClass
{
	public override string ToString()
	{
		return "MyClassStr";
	}
}

[Test]
public void String_Concatenation_with_overridden_ToString()
{
	Interpreter interpreter = new Interpreter()
		.SetVariable("myClass", new MyClass());
		
	Assert.AreEqual("ciao MyClassStr", interpreter.Eval("\"ciao \" + myClass"));
}
```

nemo048 commented:
Hello @davideicardi and @metoule 😄 Sorry for my late response. I fixed indentation (hopefully), added more tests for concatenation with null and also added concatenation tests for classes with overridden ToString() method. Could you please review my changes?

davideicardi commented on src/DynamicExpresso.Core/Parsing/Parser.cs:
```diff
@@ -2090,23 +2115,39 @@ private MethodData FindBinaryOperator(string operatorName, Expression left, Expr
+
+		private static Expression GenerateStringConcatOperand(Expression expression)
+		{
+			if (expression is ConstantExpression constantExpression &&
```
Now that you have `ToStringOrNull` function, I suspect that this `if` is not more useful. What happen if you remove it?

nemo048 replied:
Hello @davideicardi 😄 You are right. The `if` is not more required. The tests work fine without it

davideicardi commented on src/DynamicExpresso.Core/Parsing/Parser.cs:
```diff
@@ -2090,23 +2115,39 @@ private MethodData FindBinaryOperator(string operatorName, Expression left, Expr
+
+			return Expression.Condition(
+				condition,
+				Expression.ConvertChecked(nullConstant, typeof(String)),
```
Maybe here you can use directly `nullConstant`? without any conversion? At the end `null` is a valid string....

davideicardi replied:
I mean writing directly:
```
return Expression.Condition(
				condition,
				nullConstant,
				Expression.Call(expression, _toStringMethod));
```

It should be the equivalent of:
```
if (value == null) {
  return null;
}
else {
  return value.ToString();
}
```

nemo048 replied:
Unfortunately it is not possible 😞 I tried `nullConstant` without Type definition but `Expression.Condition` required that `then` and `else` if-branches should be of the same type.

davideicardi reviewed (changes requested):
Thank you!
Just two final small comments ...

nemo048 commented:
Also i found one more issue in my code. We can't directly compare value types and `null`. For example `String_Concatenation_check_string_method` test failed without current changes. Now everything seems to be fine 😄

davideicardi reviewed (approved):
Thanks! Good work!
